### PR TITLE
[Translation] XliffLintCommand: allow .xliff file extension

### PR DESCRIPTION
**Pull request by codegain**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| License       | MIT

I ran into a problem with XLIFF files having an *.xliff extension and a "target-language" attribute.
The command always outputted: There is a mismatch between the language included in the file name and the value used used in the "target-language" attribute of the file.

The "target-language" attribute was set to "en" and the filename was also "menu.en.xliff".
After reading the source code, I realized that this regex does not respect other valied file extensions such as "xliff" for these files and therefore throws this (rather confusing) error.

### CODE REVIEW

**Comment by javiereguiluz** [comment]
In other parts of the application I think we said: `xlf` is the only valid extension for XLIFF files, so we won't add support for `xliff` extension. See for example #22440.

So, in theory, we shouldn't merge this for consistency ... or we should add support for `xliff` everywhere we support `xlf`. I don't have a strong opinion about this.

**Comment by codegain** [comment]
My opinion is that if symfony is able to load .xliff files (which it is without any problems), it should be able to lint these .xliff files.

On this line (218 on master):
https://github.com/symfony/symfony/blob/7dee61ca1a4b397eef436151a3163b20eb15d47c/src/Symfony/Component/Translation/Command/XliffLintCommand.php#L218
The linter explicitly loads .xlf and .xliff files.

So this pull-request fixes a bug where xliff files are being loaded by the linter (removing that would be a BC break in my opinion), but can't be properly linted because of the wrong regex.

File: src/Symfony/Component/Translation/Command/XliffLintCommand.php [hunk 1]
```diff
@@ -126,7 +126,7 @@ private function validate($content, $file = null)
             // otherwise, both '____.locale.xlf' and 'locale.____.xlf' are allowed
             // also, the regexp matching must be case-insensitive, as defined for 'target-language' values
             // http://docs.oasis-open.org/xliff/v1.2/os/xliff-core.html#target-language
-            $expectedFilenamePattern = $this->requireStrictFileNames ? sprintf('/^.*\.(?i:%s)\.xlf/', $normalizedLocale) : sprintf('/^(.*\.(?i:%s)\.xlf|(?i:%s)\..*\.xlf)/', $normalizedLocale, $normalizedLocale);
+            $expectedFilenamePattern = $this->requireStrictFileNames ? sprintf('/^.*\.(?i:%s)\.(xlf|xliff)/', $normalizedLocale) : sprintf('/^(.*\.(?i:%s)\.(xlf|xliff)|(?i:%s)\..*\.(xlf|xliff))/', $normalizedLocale, $normalizedLocale);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Translation/Command/XliffLintCommand.php]: in strict mode, I'd suggest raising an error whe xliff is used

**Comment by nicolas-grekas** [comment]
(should target 4.4 btw)

**Comment by codegain** [comment]
> (should target 4.4 btw)

@nicolas-grekas Sorry, I'm unfamiliar with the GitHub branch targeting. I think I screwed this up >.<
I think I will open a new pull-request with your requested changes.

**Comment by codegain** [comment]
> in strict mode, I'd suggest raising an error when xliff is used

@nicolas-grekas I added another error message if "xliff" is used with strict mode enabled. Could you have a look at it again?

Altough that change does not resolve my issue, because I am using strict mode unintentionally and I found no "easy" way to disable it. The flag to enable strict mode is set to "true" by default in the constructor. I think there should be an option like "--no-strict-mode" to disable it via the command line.

**Comment by nicolas-grekas** [comment]
@codegain but why don't you use the `xlf` extension? That's the only standard one AFAIK?

**Comment by codegain** [comment]
> @codegain but why don't you use the `xlf` extension? That's the only standard one AFAIK?

@nicolas-grekas I am using poeditor.com for translations which only works with .xliff files and shows the error "File format unsupported" if an .xlf file is uploaded.

According to a section in the specs for XLIFF 1.2 (http://docs.oasis-open.org/xliff/v1.2/os/xliff-core.html):
> D.4. XLIFF File Extension
> XLIFF documents use the .xlf extension. No other extension is recommended by the specification.
the .xlf extension is recommeded, but not strictly required.

The XLIFF 2.0 specs (http://docs.oasis-open.org/xliff/xliff-core/v2.0/xliff-core-v2.0.html) make no preference towards any file extension.

**Comment by nicolas-grekas** [comment]
Did you report the issue to poeditor? That's where the real issue is to me.

**Comment by codegain** [comment]
@nicolas-grekas Their response:
> Since we support only iOS xliff files, we allow imports only with the .xliff extension, because that's the one specific to iOS.
> Uploading .xlf files may work after changing the extenstion to .xliff, but if your files are not in the same format as the iOS xliff format, data can get lost during this process.
> We have a sample of the iOS xliff format here:
> https://poeditor.com/localization/files/xliff

I had no idea that any sort of "iOS xliff" exists, because it works perfectly fine with symfony xlf files (if I change the extension to ".xliff").
I would like that this change makes into the XliffLintCommand, because my current workaround is to 
1. rename all files to ".xlf"
2. run the lint:xliff command
3. rename all files back to ".xliff"

**Comment by nicolas-grekas** [comment]
Thanks for the details. If everybody else is OK to seamlessly accept `.xliff` files, let's do it
/cc @symfony/deciders

File: src/Symfony/Component/Translation/Command/XliffLintCommand.php [hunk 2]
```diff
@@ -126,14 +126,24 @@ private function validate($content, $file = null)
...
-            $expectedFilenamePattern = $this->requireStrictFileNames ? sprintf('/^.*\.(?i:%s)\.xlf/', $normalizedLocale) : sprintf('/^(.*\.(?i:%s)\.xlf|(?i:%s)\..*\.xlf)/', $normalizedLocale, $normalizedLocale);
+            $expectedFileExtensionPattern = $this->requireStrictFileNames ? '/^.*\.xlf/' : '/^.*\.(xlf|xliff)/';
 
-            if (0 === preg_match($expectedFilenamePattern, basename($file))) {
+            if (0 === preg_match($expectedFileExtensionPattern, basename($file))) {
                 $errors[] = [
                     'line' => -1,
                     'column' => -1,
-                    'message' => sprintf('There is a mismatch between the language included in the file name ("%s") and the "%s" value used in the "target-language" attribute of the file.', basename($file), $targetLanguage),
+                    'message' => sprintf('Invalid file extension used in strict mode ("%s").', basename($file)),
```
Comment (Tobion) [inline src/Symfony/Component/Translation/Command/XliffLintCommand.php]: > in strict mode

This seems wrong. It might not be active and seems irrelevant.

File: src/Symfony/Component/Translation/Command/XliffLintCommand.php [hunk 3]
```diff
@@ -126,14 +126,24 @@ private function validate($content, $file = null)
...
                 ];
+            } else {
```
Comment (Tobion) [inline src/Symfony/Component/Translation/Command/XliffLintCommand.php]: why `else`? shouldn't you get both errors when both the extension and the content is wrong?

**Comment by codegain** [comment]
> Thanks for the details. If everybody else is OK to seamlessly accept .xliff files, let's do it

@nicolas-grekas @Tobion Just to clarify: Should the command now seamlessly accept .xliff files (without an additional error message) or raise an error if .xliff is used in strict mode?

**Comment by fabpot** [comment]
I would accept loading both without any messages.

**Comment by nicolas-grekas** [comment]
Thank you @codegain.
